mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -322,56 +322,190 @@ pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
 
         sync_status_loop_handle.notify_blocks_headers_sync_status(from_block_height + 1, to_block_height);
 
-        let mut fetch_blocker_headers_attempts = FETCH_BLOCK_HEADERS_ATTEMPTS;
+        let (block_registry, block_headers) = match retrieve_headers_helper(
+            &mut chunk_size,
+            client,
+            from_block_height + 1,
+            &mut sync_status_loop_handle,
+            ticker,
+            to_block_height,
+        )
+        .await
+        {
+            Ok(res) => res,
+            Err(_err) => break,
+        };
+
+        // Validate retrieved block headers.
+        if let Err(err) = validate_headers(ticker, from_block_height, &block_headers, storage, &spv_conf).await {
+            // This code block handles a specific error scenario where a parent hash mismatch(chain re-org) is
+            // detected in the SPV client.
+            // If this error occurs, the code retrieves and revalidates the mismatching header from the SPV client..
+            if let SPVError::ParentHashMismatch { coin, height } = &err {
+                info!("Parent hash mismatch detected for {coin} at height: {height}");
+                match retrieve_and_revalidate_mismatching_header(
+                    coin,
+                    client,
+                    *height,
+                    storage,
+                    &spv_conf,
+                    &mut sync_status_loop_handle,
+                )
+                .await
+                {
+                    Ok(_) => continue,
+                    Err(_) => break,
+                }
+            }
+
+            // Todo: remove this electrum server and use another in this case since the headers from this server are invalid
+            error!("Error {err:?} on validating the latest headers for {ticker}!");
+            sync_status_loop_handle.notify_on_permanent_error(err);
+            break;
+        };
+
+        let sleep = args.success_sleep;
+        ok_or_continue_after_sleep!(storage.add_block_headers_to_storage(block_registry).await, sleep);
+    }
+}
+
+async fn retrieve_headers_helper(
+    chunk_size: &mut u64,
+    client: &ElectrumClient,
+    from_block_height: u64,
+    sync_status_loop_handle: &mut UtxoSyncStatusLoopHandle,
```

Comment:
It's better to not pass `UtxoSyncStatusLoopHandle` here and handle permanent/temp errors notifications in the main function. Same for`retrieve_and_revalidate_mismatching_header`. We also need specific error type enum for both functions instead of a string, we can implement a function for these error types to specify if the error is permanent or not.

Reply:
done